File: 프로그래머스/소수 찾기(lv.2).java
/*
start:14:10
end:14:49
시간복잡도O(N!)
공간복잡도:O(N!)
풀이:어차피 값이 적어서 dfs로 다 돌려보면 된다!
*/
import java.util.HashSet;
import java.util.Set;
class Solution {
     Set <Integer> map=new HashSet<Integer>();
     int answer = 0;
     static int [] used=new int[10];
    boolean isTrue(int num){
        if(num==0 || num ==1)return false;
        if(num==2)return true;
        for(int i=2;i*i<=num;i++){
            if(num%i==0)return false;
        }
        return true;
    }
    void dfs(String numbers,int num){
        System.out.println(num);
        
        if(isTrue(num) && map.contains(num)==false){map.add(num);answer++;
        }
        for(int i=0;i<numbers.length();i++){
            if(used[i]==1)continue;
            used[i]=1;
            dfs(numbers,num*10+numbers.charAt(i)-'0');
            used[i]=0;
        }
    }
    public int solution(String numbers) {
        for(int i=0;i<numbers.length();i++){
            used[i]=1;
            dfs(numbers,numbers.charAt(i)-'0');
            used[i]=0;
        }
        return answer;
    }
}

File: 프로그래머스/소수찾기(lv2).cpp
/*
start:14:00
end:14:12
시간복잡도:O(N!)
공간복잡도:O(N!)
풀이:완전탐색으로 풀어주었다.
어차피 케이스가 크지않아서 충분히 가능!
*/
#include <string>
#include <vector>
using namespace std;
int vis[10]={0,};
int prime[9999999]={0,};
int answer = 0;
string number;
bool isPrime(int num){
    if(prime[num]==1)return false;
    prime[num]=1;
    if(num==1|| num==0)return false;
    if(num==2)return true;
    for(int i=2;i*i<=num;i++){
        if(num%i==0)return false;
    }
    return true;
}
void makeNum(int num){
    if(isPrime(num))answer++;
    for(int i=0;i<number.size();i++){
        if(vis[i])continue;
        vis[i]=1;
        makeNum(num*10+number[i]-'0');
        vis[i]=0;
    }
}
int solution(string numbers) {
    number=numbers;
    makeNum(0);
    return answer;
}
